src/modules/schnorr_adaptor/tests_impl.h

```diff
@@ -0,0 +1,957 @@
+    CHECK(secp256k1_memcmp_var(nonces[0], nonces[1], 32) != 0);
+}
+
+void run_adaptor_nonce_function_bip340_tests(void) {
+    unsigned char tag[20] = "SchnorrAdaptor/nonce";
+    unsigned char aux_tag[18] = "SchnorrAdaptor/aux";
+    unsigned char algo[20] = "SchnorrAdaptor/nonce";
+    size_t algolen = sizeof(algo);
+    secp256k1_sha256 sha;
+    secp256k1_sha256 sha_optimized;
+    secp256k1_scalar adaptor_scalar;
+    secp256k1_gej tj;
+    secp256k1_ge tg;
+    unsigned char nonce[32], nonce_z[32];
+    unsigned char msg[32];
+    unsigned char key[32];
+    unsigned char t[32];
+    unsigned char adaptor[33];
+    unsigned char pk[32];
+    unsigned char aux_rand[32];
+    unsigned char *args[6];
+    int i;
+    size_t size = 33;
+
+    /* Check that hash initialized by
+     * secp256k1_adaptor_nonce_function_bip340_sha256_tagged has the expected
+     * state. */
+    secp256k1_sha256_initialize_tagged(&sha, tag, sizeof(tag));
+    secp256k1_adaptor_nonce_function_bip340_sha256_tagged(&sha_optimized);
+    test_sha256_eq(&sha, &sha_optimized);
+
+   /* Check that hash initialized by
+    * secp256k1_adaptor_nonce_function_bip340_sha256_tagged_aux has the expected
+    * state. */
+    secp256k1_sha256_initialize_tagged(&sha, aux_tag, sizeof(aux_tag));
+    secp256k1_adaptor_nonce_function_bip340_sha256_tagged_aux(&sha_optimized);
+    test_sha256_eq(&sha, &sha_optimized);
+
+    secp256k1_testrand256(msg);
+    secp256k1_testrand256(key);
+    secp256k1_testrand256(t);
+    secp256k1_testrand256(pk);
+    secp256k1_testrand256(aux_rand);
+
+    secp256k1_scalar_set_b32(&adaptor_scalar, t, NULL);
+    secp256k1_ecmult_gen(&CTX->ecmult_gen_ctx, &tj, &adaptor_scalar);
+    secp256k1_ge_set_gej(&tg, &tj);
+    CHECK(secp256k1_eckey_pubkey_serialize(&tg, adaptor, &size, 1) == 1);
+
+    /* Check that a bitflip in an argument results in different nonces. */
+    args[0] = msg;
+    args[1] = key;
+    args[2] = adaptor;
+    args[3] = pk;
+    args[4] = algo;
+    args[5] = aux_rand;
+    for (i = 0; i < COUNT; i++) {
+        adaptor_nonce_function_bip340_bitflip(args, 0, 32, algolen);
+        adaptor_nonce_function_bip340_bitflip(args, 1, 32, algolen);
+        adaptor_nonce_function_bip340_bitflip(args, 2, 32, algolen);
+        adaptor_nonce_function_bip340_bitflip(args, 3, 32, algolen);
+        /* Flip algo special case "SchnorrAdaptor/nonce" */
+        adaptor_nonce_function_bip340_bitflip(args, 4, algolen, algolen);
+        /* Flip algo again */
+        adaptor_nonce_function_bip340_bitflip(args, 4, algolen, algolen);
+        adaptor_nonce_function_bip340_bitflip(args, 5, 32, algolen);
+    }
+
+    /* NULL algo is disallowed */
+    CHECK(adaptor_nonce_function_bip340(nonce, msg, key, t, pk, NULL, 0, NULL) == 0);
+    CHECK(adaptor_nonce_function_bip340(nonce, msg, key, t, pk, algo, algolen, NULL) == 1);
+    /* Other algo is fine */
```

Comment:
I don't think this test significantly increases test coverage because the nonce function does relatively little with algo. If we test algo = 0 an algo 0xff...f, why not test the same for the other nonce function arguments?

Reply:
> why not test the same for the other nonce function arguments?

I didn't consider this case. I suggested those tests because the ECDSA adaptor had them. 

I take back my suggestion. The `adaptor_nonce_function_bip340` function passes its arguments to the hash function without checking their values, except for `algo` and `aux`. Thus, a min-max test for each args seems unnecessary.